File: src/main/java/com/offz/spigot/staminaclimb/Stamina/StaminaTask.java
package com.offz.spigot.staminaclimb.Stamina;

import com.offz.spigot.staminaclimb.Climbing.ClimbBehaviour;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.GameMode;
import org.bukkit.boss.BarColor;
import org.bukkit.boss.BossBar;
import org.bukkit.entity.Player;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.util.Vector;

import java.util.UUID;

public class StaminaTask extends BukkitRunnable {
    @Override
    public void run() {
        for (UUID uuid : StaminaBar.registeredBars.keySet()) { //Regenerate stamina for every BossBar
            Player p = Bukkit.getPlayer(uuid);
            BossBar b = StaminaBar.registeredBars.get(uuid);
            double progress = b.getProgress();
            if (progress + 0.01 <= 1) {
                b.setVisible(true);
                if (p.isOnGround() || p.getGameMode().equals(GameMode.CREATIVE) || p.getVehicle() != null || !ClimbBehaviour.canClimb.get(uuid))
                    b.setProgress(progress + 0.01);
            } else {
                b.setVisible(false);
                if (p.isOnGround() || !ClimbBehaviour.canClimb.get(uuid))
                    b.setProgress(1);
            }

            if (progress <= 0.02) { //Changing bar colors and effects on player depending on its progress
                b.setColor(BarColor.RED);
                b.setTitle(ChatColor.translateAlternateColorCodes('&', "&c&lStamina")); //Make Stamina title red
                if (ClimbBehaviour.isClimbing.containsKey(uuid)) {
                    ClimbBehaviour.stopClimbing(p);
                }

                ClimbBehaviour.canClimb.put(uuid, false); //If player reaches red zone, they can't climb until they get back in green zone
                p.addPotionEffect(new PotionEffect(PotionEffectType.SLOW, 110, 2, false, false));
                p.addPotionEffect(new PotionEffect(PotionEffectType.WEAKNESS, 110, 2, false, false));
            } else if (progress < 1) {
                if (!ClimbBehaviour.canClimb.get(uuid)) { //Keep Stamina Bar red even in yellow zone while it's regenerating
                    b.setColor(BarColor.RED);
                    continue;
                }
            } else {
                b.setColor(BarColor.GREEN);
                b.setTitle(ChatColor.BOLD + "Stamina");
                ClimbBehaviour.canClimb.put(uuid, true);
            }
        }
        for (UUID uuid : ClimbBehaviour.isClimbing.keySet()) {
            //im not sure if this is needed but we're using a ConcurrentHashMap to solve some errors, so I assume it's
            //possible for the real map to not contain something we're currently iterating over
            if (!ClimbBehaviour.isClimbing.containsKey(uuid))
                return;

            Player p = Bukkit.getPlayer(uuid);
            Vector direction = p.getLocation().getDirection();
            Vector v = p.getVelocity();

            boolean isClimbing = ClimbBehaviour.isClimbing.get(uuid);

            //if climbing in creative, stop climbing but keep flight
            if(p.getGameMode().equals(GameMode.CREATIVE)) {
                ClimbBehaviour.stopClimbing(p);
                p.setAllowFlight(true);
                p.setFlying(true);
                continue;
            }

            if (!p.isFlying() && isClimbing || p.getFallDistance() > 5) {
                ClimbBehaviour.stopClimbing(p);
                continue;
            }

            if (ClimbBehaviour.atWall(p.getLocation(), uuid)) {
                if (ClimbBehaviour.cooldown.get(uuid) <= System.currentTimeMillis())
                    ClimbBehaviour.cooldown.put(uuid, System.currentTimeMillis());
                p.setAllowFlight(true);
                p.setFlying(true);
                ClimbBehaviour.isClimbing.put(uuid, true);
            } else {
                if (isClimbing) {
                    p.setVelocity(new Vector(v.getX() + direction.getX() / 20, v.getY(), v.getZ() + direction.getZ() / 20));
                    ClimbBehaviour.isClimbing.put(uuid, false);
                    ClimbBehaviour.cooldown.put(uuid, System.currentTimeMillis());
                    p.setFlying(false);
                    p.setAllowFlight(false);
                } else if (ClimbBehaviour.cooldown.get(uuid) + 400 <= System.currentTimeMillis()) {
                    ClimbBehaviour.stopClimbing(p);
                    continue;
                }
            }
            if (ClimbBehaviour.isClimbing.get(uuid))
                StaminaBar.removeProgress(0.008, uuid);
        }
    }
}

File: src/main/java/com/offz/spigot/staminaclimb/stamina/StaminaTask.kt
package com.offz.spigot.staminaclimb.stamina

import com.mineinabyss.idofront.messaging.color
import com.offz.spigot.staminaclimb.*
import com.offz.spigot.staminaclimb.climbing.ClimbBehaviour
import com.offz.spigot.staminaclimb.config.StaminaConfig
import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.util.Vector
import org.bukkit.Location
import org.bukkit.entity.Player
import org.bukkit.boss.BarColor
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import org.bukkit.scheduler.BukkitRunnable
import java.util.Calendar
import java.util.UUID
import kotlin.math.abs

class StaminaTask : BukkitRunnable() {

    private var previousTickTime: Long = 0;
    private val targetDeltaTime = 1.0 / 20.0

    // Not sure how to get the World to make a Location so I'm just doing this | TODO: change to a Location
    private var prevX = 0.0
    private var prevY = 0.0
    private var prevZ = 0.0

    private var ticksSinceLastColorSwitch = 0
    private val ticksToSwitchColor = 5

    private var currentWallMultiplier = 1.0

    internal fun isPlayerMoving(player: Player): Boolean {
        val minMovementValue = 0.007
        val isMoving = abs(player.getLocation().getX() - prevX) >= minMovementValue ||
                       abs(player.getLocation().getY() - prevY) >= minMovementValue ||
                       abs(player.getLocation().getZ() - prevZ) >= minMovementValue 
        prevX = player.getLocation().getX()
        prevY = player.getLocation().getY()
        prevZ = player.getLocation().getZ()
        return isMoving
    }

    override fun run() {
        StaminaBar.registeredBars.keys.forEach { uuid ->
            val player = Bukkit.getPlayer(uuid) ?: StaminaBar.registeredBars.remove(uuid).let { return@forEach }
            val bar = StaminaBar.registeredBars[uuid] ?: return@forEach
            val progress = bar.progress

            bar.isVisible = progress + StaminaConfig.data.staminaRegen <= StaminaConfig.data.barMax //hide when full

            if (!uuid.isClimbing) { //Regenerate stamina
                if(player.isOnGround)
                    StaminaBar.addProgressWithDeltaTime(StaminaConfig.data.staminaRegen
                        .coerceAtMost(StaminaConfig.data.barMax), uuid)
                else 
                    StaminaBar.addProgressWithDeltaTime(StaminaConfig.data.staminaRegenInAir
                        .coerceAtMost(StaminaConfig.data.barMax), uuid)
            } else { //Lose stamina
                if(isPlayerMoving(player)) {
                    StaminaBar.removeProgressWithDeltaTime(StaminaConfig.data.staminaRemoveWhileMoving * currentWallMultiplier, uuid)
                }
                else {
                    StaminaBar.removeProgressWithDeltaTime(StaminaConfig.data.staminaRemovePerTick * currentWallMultiplier, uuid)
                }
            }

            if (progress <= StaminaConfig.data.barMin) { //Changing bar colors and effects on player depending on its progress
                bar.color = BarColor.RED
                bar.setTitle("&c&lStamina".color()) //Make Stamina title red
                if (uuid.isClimbing) {
                    player.stopClimbing()
                    uuid.canClimb = false //If player reaches red zone, they can't climb until they get back in green zone
                    player.addPotionEffect(PotionEffect(PotionEffectType.SLOW, 110, 2, false, false))
                    player.addPotionEffect(PotionEffect(PotionEffectType.WEAKNESS, 110, 2, false, false))
                }
            } else if (progress < StaminaConfig.data.barMax && !uuid.canClimb) {
                bar.color = BarColor.RED //Keep Stamina Bar red even in yellow zone while it's regenerating
            } else if (uuid.isClimbing && progress < StaminaConfig.data.barFlashPoint / 2.0) {
                ticksSinceLastColorSwitch++
                if (ticksSinceLastColorSwitch >= ticksToSwitchColor / 2) {
                    if(bar.color == BarColor.RED)
                        bar.color = BarColor.YELLOW
                    else
                        bar.color = BarColor.RED
                    ticksSinceLastColorSwitch = 0
                }
            } else if (uuid.isClimbing && progress < StaminaConfig.data.barFlashPoint) {
                ticksSinceLastColorSwitch++
                if (ticksSinceLastColorSwitch >= ticksToSwitchColor) {
                    if(bar.color == BarColor.RED)
                        bar.color = BarColor.GREEN
                    else
                        bar.color = BarColor.RED
                    ticksSinceLastColorSwitch = 0
                }
            } else {
                bar.color = BarColor.GREEN
                bar.setTitle("&lStamina".color())
                uuid.canClimb = true
            }
        }

        // --------------------------------------

        var specialWall = false

        ClimbBehaviour.isClimbing.entries.forEach { (uuid, isClimbing) ->
            val player = Bukkit.getPlayer(uuid) ?: ClimbBehaviour.isClimbing.remove(uuid).let { return }

            //if climbing in creative, stop climbing but keep flight
            if (player.gameMode == GameMode.CREATIVE) {
                player.stopClimbing()
                player.allowFlight = true
                player.isFlying = true
                return@forEach
            }

            //prevent player from climbing if they have fallen far enough TODO dunno what hte flying check is for
            if (!player.isFlying && isClimbing || player.fallDistance > StaminaConfig.data.maxFallDist) {
                player.stopClimbing()
                return@forEach
            }
            val atWallMultiplier = player.wallDifficulty
            if (atWallMultiplier >= 0) {
                if (uuid.climbCooldownDone) uuid.restartCooldown()
                player.allowFlight = true
                player.isFlying = true
                ClimbBehaviour.isClimbing[uuid] = true
            } else {
                if (isClimbing) {
                    uuid.isClimbing = false
                    uuid.restartCooldown()
                    player.launchInDirection()
                    player.isFlying = false
                    player.allowFlight = false
                }
                //only prevent air jump after AIR_TIME ms
                else if (uuid.climbCooldown + StaminaConfig.data.airTime < 0) {
                    player.stopClimbing()
                    return@forEach
                }
            }

            //if (isClimbing) StaminaBar.removeProgressWithDeltaTime(StaminaConfig.data.staminaRemovePerTick * atWallMultiplier, uuid)
            specialWall = true
            currentWallMultiplier = atWallMultiplier
        }

        if (!specialWall)
            currentWallMultiplier = 1.0
    }
}
